[#17] 좋아요 기능을 레디스로 캐시

jzakka opened the pull request:
### 기능
1. 사용자가 like 요청
2-1. redis에 이미 데이터 있니? yes -> block, no -> 2-2
2-2. db에 이미 데이터 있니? yes -> block, no -> 3
3. redis에 데이터 저장
4. 스케줄러가 주기적으로 redis를 DB로 업데이트

- 캐싱되는 데이터는 
feed_id : {liker1_id, liker2_id, ..}와 같이 피드와 그에 따른 좋아요한 사람의 id가 셋으로 저장되는 구조
- DB에 쓰기는 스케줄러에 의해서만 동작

jzakka committed:
Refactor: LikeRedisRepository를 LikeService에서 직접 사용하도록 수정

jzakka committed:
Test: 좋아요 서비스 캐싱로직 테스트 추가

jzakka committed:
Feat: 좋아요 캐싱
- LikeRedisRepository 추가

f-lab-moony commented on src/main/java/com/stoury/config/RedisConfig.java:
```diff
@@ -0,0 +1,26 @@
+public class RedisConfig {
+    @Bean
+    public RedisConnectionFactory redisConnectionFactory() {
+        LettuceConnectionFactory connectionFactory = new LettuceConnectionFactory("127.0.0.1", 6379);
```
접속정보는 따로 설정으로 빼는게 좋을 것 같아요 ~

f-lab-moony commented on src/main/java/com/stoury/repository/LikeRedisRepository.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.Set;
+
+@Repository
+@RequiredArgsConstructor
```
요건 사용을 아예 안하는게 좋을 것 같아요 ~ 생성자로 빼주시면 아래 `@PostConstruct`도 필요 없지 않을까요 ?

f-lab-moony commented on src/main/java/com/stoury/repository/LikeRedisRepository.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    public Like save(Like like) {
+        String feedId = Objects.requireNonNull(like.getFeed().getId(), "Feed id cannot be null").toString();
```
유효성 검사와 실제 사용하는 부분은 분리가 되면 좋겠어요. 어떤 값을 쓰는지 잘 안보여요 ~

f-lab-moony commented on src/main/java/com/stoury/repository/LikeRedisRepository.java:
```diff
@@ -0,0 +1,57 @@
+        return false;
+    }
+
+    public int countByFeed(Feed feed) {
```
요거 그냥 반환값을 long으로 바꾸는게 낫지 않을까요 ?

f-lab-moony commented on src/main/java/com/stoury/service/CommentService.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Transactional(readOnly = true)
+    public List<ChildCommentResponse> getChildComments(Long parentCommentId, LocalDateTime orderThan) {
```
요거 약간 궁금한데, 대댓글도 보통 페이징을 하나요? 대댓글은 한번에 쫙 펴졌던거같은데 ..

jzakka replied:
서비스마다 다 다르더라구요.
유튜브나 인스타는 대댓글도 페이징이 됐었고, 네이버 카페는 댓글 대댓글이 전부 한번에 나왔던 것 같습니다.

f-lab-moony replied:
오 그렇군요 ? 워낙 사용자가 많으면 감당이 안돼서 그런가 ㅋㅋ 넵 좋습니다~

f-lab-moony commented on src/main/java/com/stoury/service/LikeService.java:
```diff
@@ -18,37 +19,45 @@
+                .orElseThrow(FeedSearchException::new);
 
-        if (likeRepository.existsByMemberAndFeed(liker, feed)) {
+        if (likeRedisRepository.existsByMemberAndFeed(liker, feed) || likeJpaRepository.existsByMemberAndFeed(liker, feed)) {
```
요거 레디스만 봐도 되지 않을까요 ? 구조상 레디스가 원천데이터고 db는 백업이라 레디스에 없는데 db에 있는 경우는 없을 것 같아서요 ~

jzakka replied:
앗, 저는 레디스를 캐싱용으로 사용하려고 캐시 일관성 때문에 둘다 확인을 했었습니다.

만약 사용자가 좋아요를 하고 스케줄러가 DB에 좋아요를 업데이트하면,
레디스에는 해당 좋아요가 없어지고 이제 DB에 남게됩니다.

이 상황에서 레디스만 체크하는 구조라면, 사용자가 다시 좋아요를 누르면 레디스에는 데이터가 없으므로 중복 좋아요가 가능해지기 때문에 DB도 체크를 하도록 구현했습니다.

<img width="1420" alt="Cache" src="https://github.com/f-lab-edu/Stoury/assets/105845911/5f7dc35e-bca6-4ff7-9c99-3cfc87bcf31e">


근데 스케줄러가 중복 좋아요를 DB로 업데이트하려 해도 Unique constraint 때문에 중복 좋아요로 계산되진 않겠네요. 그럼 굳이 DB접속을 할 필요 없이 redis만 체크해도 괜찮을 것 같긴 합니다. 근데 DB에서 일어날 수 있는 예외는 애플리케이션에서 미리 검증하라고 하신 피드백이 생각나서...  
이 부분은 어떻게 하면 좋을까요?😅

f-lab-moony replied:
아하 .. 근데 그러면 레디스의 캐시로서의 효율이 떨어질 것 같아요. 제가 생각했던 건 레디스를 메모리 캐시로 사용하고 주기적으로 db에 싱크를 맞추는걸 생각했었어요. 약간 레디스 영구저장을 직접 구현하는 느낌이긴 한데.. 아니면 rdb와 관계를 아예 끊고 [이런 걸](https://inpa.tistory.com/entry/REDIS-%F0%9F%93%9A-%EB%8D%B0%EC%9D%B4%ED%84%B0-%EC%98%81%EA%B5%AC-%EC%A0%80%EC%9E%A5%ED%95%98%EB%8A%94-%EB%B0%A9%EB%B2%95-%EB%8D%B0%EC%9D%B4%ED%84%B0%EC%9D%98-%EC%98%81%EC%86%8D%EC%84%B1) 이용해봐도 좋을 것 같다고 생각합니다 ~

f-lab-moony commented on src/main/java/com/stoury/repository/LikeRedisRepository.java:
```diff
@@ -0,0 +1,57 @@
+        return Math.toIntExact(opsForSet.size(feedId));
+    }
+
+    private void clearAllOnlyForTest() {
```
요건 어디서 쓰는거죠 ?

jzakka replied:
테스트에서 실제로 레디스에 데이터가 저장되는지를 확인합니다.
매 테스트마다 데이터를 삭제해주려고 만들었습니다.

f-lab-moony replied:
private이긴 하지만 서비스에 필요 없는 걸 테스트를 위해 서비스에 구현하는게 별로 적절해보이진 않는데요, 클리어하는 기능을 따로 빼서 테스트의 beforEach 같은데에 수행해주는게 좋지 않을까요 ?

f-lab-moony commented on src/main/java/com/stoury/service/LikeService.java:
```diff
@@ -18,37 +19,45 @@
+        if (likeRedisRepository.deleteByMemberAndFeed(liker, feed)) {
+            return;
+        }
+        likeJpaRepository.deleteByMemberAndFeed(liker, feed);
```
delete도 매번 날아가는건 지양하는게 좋을 것 같아요. 그리고 레디스에 넣는거만 있고 db에 넣는 작업이 안보이는거같은데, 스케줄링 하시는걸로 알고 있는데 아직 작업이 안된걸까요 ~? 스케줄링 된 이후에도, save가 곧바로 이루어지는게 아니라서 지금 상태라면 delete랑 동기화가 안될 것 같아 보여요 ~

jzakka replied:
스케줄링은 아직 안 만들었습니다. 따로 이슈생성하고 만드는게 왠지 보기 편하실 것 같아서...

delete도 피드백대로 수정해보겠습니다😃

f-lab-moony reviewed (changes requested):
고생 많으셨습니다 ~
피드백 확인 부탁드려요!

jzakka committed:
Refactor: 레디스 접속 팩토리를 설정파일로 옮김

jzakka committed:
Refactor: Redis리포지토리의 opsForSet을 생성자가 만들도록 함

jzakka committed:
Refactor: 유효성 검증을 메서드 추출

jzakka committed:
Refactor: 좋아요수 long으로 반환

jzakka committed:
Refactor: LikeRedisRepository에서 테스트용 메서드 삭제

jzakka committed:
Refactor: redis가 원천데이터 소스가 되도록 수정

- jpa를 사용하던 리포지토리 삭제
- Like테이블 삭제
- LikeRedisRepository.java -> LikeRedisRepository.java

f-lab-moony reviewed (approved):
고생 많으셨습니다! 바로 머지 하도록 할게요 ~
